src/pages/index.js

```diff
@@ -32,23 +42,120 @@ addPostButton.addEventListener("click", () => {
   postFormPopup.open();
 });
 
+//Handle Submit functions
+function handleSubmit(request, popupModal, popupForm, renderingText, postText) {
+  renderLoading(true, popupForm, renderingText, postText);
+  request()
+    .then(() => {
+      popupModal.close();
+    })
+    .catch(console.error)
+    .finally(() => {
+      renderLoading(false, popupForm, renderingText, postText);
+    });
+}
+
+const handleProfilePictureSubmit = ({ picture }) => {
+  function makeRequest() {
+    return api.updateAvatar(picture).then((profileInfo) => {
+      profilePicture.src = profileInfo.avatar;
+    });
+  }
+
+  handleSubmit(makeRequest, profilePictureFormPopup, editProfilePictureForm);
+};
+
 const handleProfileFormSubmit = ({ name, job }) => {
-  userInfo.setUserInfo({ name, job });
-  profileFormPopup.close();
+  function makeRequest() {
+    return api.updateProfileInfo({ name, about: job }).then((newData) => {
+      userInfo.setUserInfo({
+        name: newData.name,
+        job: newData.about,
+      });
+    });
+  }
+
+  handleSubmit(makeRequest, profileFormPopup, editProfileForm);
 };
 
 const handlePostFormSubmit = ({ title, url }) => {
   const newPost = {
     name: title,
     link: url,
   };
-  const newCard = createCard(newPost);
-  cardSection.addNewItem(newCard);
-  postFormPopup.close();
+
+  function makeRequest() {
+    return api.createApiCard(newPost).then((createdCard) => {
+      cardSection.addNewItem(createCard(createdCard));
+    });
+  }
+
+  handleSubmit(makeRequest, postFormPopup, addPostForm);
 
   formValidators[selectors.addPostForm].disableButton();
 };
 
+const handleDelete = (card) => {
+  confirmDeletePopup.open();
+  formValidators[selectors.confirmDeleteForm].enableButton();
+  confirmDeletePopup.setSubmitFunction(() => {
+    function makeRequest() {
+      return api.deleteCard(card._data._id).then(() => {
+        card.handleDeletePost();
+      });
+    }
+
+    handleSubmit(
+      makeRequest,
+      confirmDeletePopup,
+      confirmDeleteForm,
+      "Deleting...",
+      "Delete"
+    );
+  });
+};
+
+const handleLike = (card) => {
+  const updateLikeState = () => {
+    card._data.isLiked = !card._data.isLiked;
+  };
+
+  if (!card._data.isLiked) {
+    api
+      .likeCard(card._data._id)
+      .then(() => {
+        card.handleAddLikeButton();
+        updateLikeState();
+      })
+      .catch((error) => {
+        console.error(`Error adding like: ${error}`);
+      });
+  } else {
+    api
+      .dislikeCard(card._data._id)
+      .then(() => {
+        card.handleRemoveLikeButton();
+        updateLikeState();
+      })
+      .catch((error) => {
+        console.error(`Error removing like: ${error}`);
+      });
+  }
+};
+
+function renderLoading(
+  isLoading,
+  modal,
+  renderingText = "Saving...",
+  postText = "Save"
+) {
+  if (isLoading) {
+    modal.querySelector(".modal__button").textContent = renderingText;
+  } else {
+    modal.querySelector(".modal__button").textContent = postText;
```

Comment:
**Could be improved**

It’s better to make a `renderLoading` method inside `PopupWithForm`. You should find the button once in the constructor and change the text:  

```js
  constructor({ popupSelector, handleSubmit }) {
    super(popupSelector);
    // fix the initial button text only once in the constructor
    this._submitBtnText = this._submitBtn.textContent
  }
  // add 2 params: isLoading and loadingText with a default text
  renderLoading(isLoading, loadingText='Saving...') {
    if (isLoading) {
      this._submitBtn.textContent = loadingText;
    } else {
	// here we return back the initial text. So, you don’t need to bother yourself about it
      this._submitBtn.textContent = this._submitBtnText;
    }
  }

```